certificate/endpoint.go

```diff
@@ -25,7 +27,15 @@ type Endpoint struct {
 
 // NewEndpoint creates a new certificate signing gRPC server.
 func NewEndpoint(log *zap.Logger, ca *identity.FullCertificateAuthority, authorizationDB *authorization.DB, minDifficulty uint16) *Endpoint {
+	codeMap := errs2.CodeMap{
+		&authorization.ErrNotFound: rpcstatus.Unauthenticated,
+		&authorization.ErrInvalidClaim: rpcstatus.InvalidArgument,
+		&authorization.ErrAlreadyClaimed: rpcstatus.ResourceExhausted,
```

Comment:
I hesitate that this error should be mapped to `RESOURCE_EXAUSTED` code, it looks for me more an `INVALID_ARGUMENT`.

Reply:
I disagree with `INVALID_ARGUMENT` being appropriate as it's described by [the docs]() as:
>indicat[ing] arguments that are problematic regardless of the state of the system...
I would argue that the state of the system is what makes this error possible; an authorization exists but has been claimed already. It's noteworthy that authorizations can be unclaimed by administrators of this system.

`PERMISSION_DENIED` might seem like a reasonable choice as well but the docs describe the former as so:
>The caller does not have permission to execute the specified operation. PERMISSION_DENIED must not be used for rejections caused by exhausting some resource (use RESOURCE_EXHAUSTED instead for those errors). PERMISSION_DENIED must not be used if the caller can not be identified (use UNAUTHENTICATED instead for those errors). This error code does not imply the request is valid or the requested entity exists or satisfies other pre-conditions.

In this case we **can** imply that the request is valid and that it would otherwise succeed by using `RESOURCE_EXHAUSTED`. `UNAUTHENTICATED`; therefore, also seems inappropriate because we **can** identify the user.

Reply:
You're right regarding `INVALID_ARGUMENT`, I agree of not being `UNAUTHENTICATED`, but I'm not sold about using `RESOURCE_EXHAUSTED`, but OK.

Reply:
I'm don't see what option would apply better than `RESOURCE_EXHAUSTED`.

Reply:
For HTTP usually https://httpstatuses.com/409 is submitted. Also https://cloud.google.com/apis/design/errors

Reply:
This is rpc though; @egonelbre what do you suggest?

Reply:
Aborted?

Reply:
I think that`FAILED_PRECONDITION` could be an option because client can success using a not claimed one but because you commented

> It's noteworthy that authorizations can be unclaimed by administrators of this system.

Then, I have doubts on it.

Reply:
`ABORTED`, `FAILED_PRECONDITION`, `ALREADY_EXISTS` all sound plausible.

There are quite a few more documents status codes how they are used. e.g. https://developers.google.com/maps-booking/reference/grpc-api/status_codes

But, `ALREADY_EXISTS` sounds to me the best choice. `FAILED_PRECONDITION` sounds like the system can be adjusted to fix the issue, `ABORTED` sounds like there can be something fixed. `ALREADY_EXISTS` seems to indicate, that something has been already created and you cannot redo that operation.